xonsh/dirstack.py

```diff
@@ -41,7 +41,7 @@ def _try_cdpath(apath):
     env = builtins.__xonsh_env__
     cdpaths = env.get('CDPATH')
     for cdp in cdpaths:
-        for cdpath_prefixed_path in iglob(os.path.join(cdp, apath)):
+        for cdpath_prefixed_path in iglob(os.path.expanduser(os.path.join(cdp, apath))):
```

Comment:
Perhaps you could change `os.path.expanduser()` to `builtins.__xonsh_expand_path__()` (which you can also see as `xonsh.built_ins.expand_path()`) so that this works for environment variable names too, like `${HOME}/mydir`.